src/wallet.rs

```diff
@@ -0,0 +1,219 @@
+    #[error(transparent)]
+    MetadataManager(#[from] MetadataManagerError),
+    #[error(transparent)]
+    KeyManager(#[from] KeyManagerError),
+    #[error("Required credential not found for input descriptor id: {0}")]
+    RequiredCredentialNotFound(String),
+    #[error("Key not found for index: {0}")]
+    KeyNotFound(u8),
+    #[error("Failed to deserialize JSON: {0}")]
+    Deserialization(String),
+    #[error("Serde JSON Error: {0}")]
+    SerdeJson(String),
+    #[error("Failed to submit OID4VP response: {0}")]
+    OID4VPResponseSubmission(String),
+    #[error("Failed to create presentation submission: {0}")]
+    PresentationSubmissionCreation(String),
+    #[error("Failed to create presentation: {0}")]
+    VerifiablePresentation(String),
+    #[error("Failed to create JWT: {0}")]
+    GenerateJwt(String),
+    #[error("Failed to acquire read/write lock for active key index: {0}")]
+    ActiveKeyIndexReadWriteError(String),
+    #[error("Failed to parse JSON encoded JWK: {0}")]
+    JWKParseError(String),
+    #[error("Failed to generate DID Key from JWK: {0}")]
+    DIDKeyGenerateUrl(String),
+    #[error("Invalid DID URL: {0}")]
+    InvalidDIDUrl(String),
+    #[error("Unsupported Response Mode for OID4VP Request: {0}")]
+    OID4VPUnsupportedResponseMode(String),
+    #[error("Signing Algorithm Not Found: {0}")]
+    SigningAlgorithmNotFound(String),
+    #[error("Failed to sign JWT: {0}")]
+    SigningError(String),
+    #[error("Failed to save reference to credential.")]
+    InvalidCredentialReference,
+    #[error(transparent)]
+    CredentialCallback(#[from] CredentialCallbackError),
+    #[error("Unknown error occurred.")]
+    Unknown,
+}
+
+// Handle unexpected errors when calling a foreign callback
+impl From<uniffi::UnexpectedUniFFICallbackError> for WalletError {
+    fn from(value: uniffi::UnexpectedUniFFICallbackError) -> Self {
+        WalletError::UnexpectedUniFFICallbackError(value.reason)
+    }
+}
+
+/// The [Wallet](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#section-2-3.32) is used by the
+/// Holder to receive, store, present, and manage
+/// [Verifiable Credentials](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#section-2-3.4) and key material.
+/// There is no single deployment model of a Wallet:
+/// [Verifiable Credentials](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#section-2-3.4) and keys can
+/// both be stored/managed locally, or by using a remote self-hosted service, or a remote third-party service.
+///
+/// In the context of the [OID4VP](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html) specification,
+/// the Wallet acts as an OAuth 2.0 Authorization Server
+/// (see [RFC6749](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#RFC6749))
+/// towards the Credential Verifier which acts as the OAuth 2.0 Client.
+///
+#[derive(uniffi::Object, Debug)]
+pub struct Wallet {
```

Comment:
TODO: Rename `Wallet` to `SpruceKit` or `SpruceSDK` to keep it generic for top-level mobile-sdk-rs driver. This falls in alignment with the renaming of the `wallet-sdk` -> `mobile-sdk`.

Reply:
FYI wallet-sdk was renamed to mobile-sdk because we want the SDK to encapsulate more than just wallet capabilities. So I don't think that's enough of a reason to rename it.

Reply:
Save for a later PR.